worldedit-core/src/main/java/com/fastasyncworldedit/core/extent/processor/heightmap/HeightmapProcessor.java

```diff
@@ -119,6 +121,66 @@ public IChunkSet processSet(IChunk chunk, IChunkGet get, IChunkSet set) {
                     }
                 }
             }
+        } else {
+                int[] setSection = hasSectionSet ? set.loadIntsIfPresent(layer) : null;
+                if (setSection == null || Arrays.equals(setSection, FaweCache.INSTANCE.EMPTY_INT_4096) ||
+                        Arrays.equals(setSection, AIR_LAYER_INT)) {
+                    hasSectionSet = false;
+                }
+                if (!hasSectionSet && !hasSectionGet) {
+                    continue;
+                }
+                int[] getSection = null;
+                for (int y = 15; y >= 0; y--) {
+                    // We don't need to actually iterate over x and z as we're both reading and writing an index
+                    for (int j = 0; j < BLOCKS_PER_Y; j++) {
+                        int ordinal = 0;
+                        if (hasSectionSet) {
+                            ordinal = setSection[index(y, j)];
+                        }
+                        if (ordinal == 0) {
+                            if (!hasSectionGet) {
+                                if (!hasSectionSet) {
+                                    continue layer;
+                                }
+                                continue;
+                            } else if (getSection == null) {
+                                getSection = get.loadInts(layer);
+                                // skip empty layer
+                                if (Arrays.equals(getSection, FaweCache.INSTANCE.EMPTY_INT_4096)
+                                        || Arrays.equals(getSection, AIR_LAYER_INT)) {
+                                    hasSectionGet = false;
+                                    if (!hasSectionSet) {
+                                        continue layer;
+                                    }
+                                    continue;
+                                }
+                            }
+                            ordinal = getSection[index(y, j)];
+                        }
+                        // fast skip if block isn't relevant for any height map (air or empty)
+                        if (ordinal < 4) {
```

Comment:
Please refactor, given this is copy/pasted from the `if` block.

Reply:
I have refactored this into a private method with my next commit.